src/ocamlorg/lib/package.ml

```diff
@@ -58,58 +103,177 @@ module Info = struct
     let data = OpamFile.OPAM.depopts opam in
     parse_formula data
 
-  let of_opamfile (opam : OpamFile.OPAM.t) =
+  module Lwt_fold = struct
+    let fold fn =
+      let open Lwt.Syntax in
+      Lwt_list.fold_left_s (fun acc (key, value) ->
+          let* () = Lwt.pause () in
+          fn key value acc)
+
+    let package_map fn map init = OpamPackage.Map.bindings map |> fold fn init
+
+    let package_name_map fn map init =
+      OpamPackage.Name.Map.bindings map |> fold fn init
+
+    let package_version_map fn map init =
+      OpamPackage.Version.Map.bindings map |> fold fn init
+  end
+
+  let rev_depends deps =
+    Lwt_fold.package_map
+      (fun pkg version acc ->
+        Lwt.return
+        @@ OpamPackage.Set.fold
+             (fun pkg1 ->
+               OpamPackage.Map.update
+                 pkg1
+                 (OpamPackage.Set.add pkg)
+                 OpamPackage.Set.empty)
+             version
+             acc)
+      deps
+      OpamPackage.Map.empty
+
+  let mk_revdeps pkg pkgs rdepends =
+    let open OpamTypes in
+    Lwt_fold.package_name_map
+      (fun name versions acc ->
+        let vf =
+          OpamFormula.formula_of_version_set
+            (OpamPackage.versions_of_name pkgs name)
+            versions
+        in
+        let _latest =
```

Comment:
We should probably expose this `latest` to the code, this is useful for linking correctly i.e. a package that has a dependency like `A = version` currently just links to the latest version of the package which is wrong. We could do it also in a separate PR, so feels free to remove this :))

Reply:
For now I prefer to focus on what we have which is already quite a chunk (marshal state / switch to lwt for heavy computations / rev deps computation and display).
Let's do that in a separate PR !